code/cabal-audit/src/Distribution/Audit.hs

```diff
@@ -0,0 +1,247 @@
+        , show errs
+        ]
+    CabalException ctx (SomeException ex) ->
+      "cabal failed while "
+        <> ctx
+        <> ":\n"
+        <> displayException ex
+
+-- | the type of output that is chosen for the command
+data Output
+  = -- | write humand readable to stdout
+    HumanReadable
+  | -- | write as Osv format to the specified file
+    Osv {osvPath :: FilePath}
+
+-- | configuration that is specific to the cabal audit command
+data AuditConfig = MkAuditConfig
+  { advisoriesPathOrURL :: Either FilePath String
+  -- ^ path or URL to the advisories
+  , verbosity :: Verbosity.Verbosity
+  -- ^ verbosity of cabal
+  , outputType :: Output
+  }
+
+-- | the main action to invoke
+auditMain :: IO ()
+auditMain =
+  do
+    (auditConfig, nixStyleFlags) <- customExecParser (prefs showHelpOnEmpty) do
+      info (helper <*> auditCommandParser) do
+        mconcat
+          [ fullDesc
+          , progDesc (formatWith [blue] "audit your cabal projects for vulnerabilities")
+          , header (formatWith [bold, blue] "Welcome to cabal audit")
+          ]
+
+    buildAdvisories auditConfig nixStyleFlags
+      >>= handleBuiltAdvisories (outputType auditConfig)
+    `catch` \(SomeException ex) -> do
+      hPutStrLn stderr $
+        unlines
+          [ formatWith [red, bold] "cabal-audit failed:"
+          , formatWith [red] $ displayException ex
+          ]
+      exitFailure
+
+buildAdvisories :: AuditConfig -> NixStyleFlags () -> IO (M.Map PackageName ElaboratedPackageInfoAdvised)
+buildAdvisories MkAuditConfig {advisoriesPathOrURL, verbosity} flags = do
+  let cliConfig = projectConfigFromFlags flags
+
+  ProjectBaseContext {distDirLayout, cabalDirLayout, projectConfig, localPackages} <-
+    establishProjectBaseContext
+      verbosity
+      cliConfig
+      OtherCommand
+      `catch` \ex -> throwIO $ CabalException {reason = "trying to establish project base context", cabalException = ex}
+  -- the two plans are
+  -- 1. the "improved plan" with packages replaced by in-store packages
+  -- 2. the "original" elaborated plan
+  --
+  -- as far as I can tell, for our use case these should be indistinguishable
+  (_improvedPlan, plan, _, _, _) <-
+    rebuildInstallPlan verbosity distDirLayout cabalDirLayout projectConfig localPackages Nothing
+      `catch` \ex -> throwIO $ CabalException {reason = "elaborating the install-plan", cabalException = ex}
+
+  when (verbosity > Verbosity.normal) do
+    hPutStrLn stderr (formatWith [blue] "Finished building the cabal install plan, looking for advisories...")
+
+  advisories <- do
+    let k realPath =
+          listAdvisories realPath
+            >>= validation (throwIO . ListAdvisoryValidationError realPath) pure
+    case advisoriesPathOrURL of
+      Left fp -> k fp
+      Right url -> withSystemTempDirectory "cabal-audit" \tmp -> do
+        hPutStrLn stderr $ formatWith [blue] $ "trying to clone " <> url
+        callProcess "git" ["clone","--depth", "1",url, tmp]
```

Comment:
Also under "future work" probably: passing a subdir to descend into before looking for the "advisories" directory.

Reply:
good idea! I think the "looking for advisories" directory should in itself be refactored, but I will keep this in mind.